platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/tipset/TipsetScoreCalculator.java

```diff
@@ -0,0 +1,254 @@
+            throw new IllegalArgumentException("event creator must be the same as the window ID");
+        }
+
+        final Tipset eventTipset = tipsetBuilder.getTipset(event);
+        if (eventTipset == null) {
+            throw new IllegalArgumentException("event " + event + " is not in the tipset tracker");
+        }
+
+        final long score = snapshot.getWeightedAdvancementCount(selfId, eventTipset);
+        if (score > maximumPossibleScore) {
+            throw new IllegalStateException(
+                    "score " + score + " is greater than the maximum possible score " + maximumPossibleScore);
+        }
+
+        final long scoreImprovement = score - previousScore;
+
+        if (isSuperMajority(score + selfWeight, totalWeight)) {
+            snapshot = eventTipset;
+            snapshotHistory.add(snapshot);
+            if (snapshotHistory.size() > snapshotHistorySize) {
+                snapshotHistory.remove();
+            }
+            previousScore = 0;
+        } else {
+            previousScore = score;
+        }
+
+        return scoreImprovement;
+    }
+
+    /**
+     * Figure out what advancement score we would get if we created an event with a given list of parents.
+     *
+     * @param parents the proposed parents of an event
+     * @return the advancement score we would get by creating an event with the given parents
+     */
+    public long getTheoreticalAdvancementScore(@NonNull final List<EventDescriptor> parents) {
+        if (parents.isEmpty()) {
+            return 0;
+        }
+
+        final List<Tipset> parentTipsets = new ArrayList<>(parents.size());
+        for (final EventDescriptor parent : parents) {
+            parentTipsets.add(tipsetBuilder.getTipset(parent));
+        }
+
+        // Don't bother advancing the self generation, since self advancement doesn't contribute to tipset score.
+        final Tipset newTipset = Tipset.merge(parentTipsets);
+
+        return snapshot.getWeightedAdvancementCount(selfId, newTipset) - previousScore;
+    }
+
+    /**
+     * Compute the current maximum bully score with respect to all nodes. This is a measure of how well slow node's
+     * events are being incorporated in the hashgraph by faster nodes. A high score means slow nodes are being bullied
+     * by fast nodes. A low score means slow nodes are being included in consensus. Lower scores are better.
+     *
+     * @return the current tipset bully score
+     */
+    public int getBullyScore() {
+        int bullyScore = 0;
+        for (int nodeIndex = 0; nodeIndex < nodeCount; nodeIndex++) {
+            bullyScore = Math.max(bullyScore, getBullyScoreForNodeIndex(nodeIndex));
+        }
+        return bullyScore;
+    }
+
+    /**
+     * Get the bully score with respect to one node. A high bully score means that we have access to events that could
+     * go into our ancestry, but for whatever reason we have decided not to put into our ancestry.
```

Comment:
This doc describes what a *high* bully score means, but not exactly what a given bully score maps to conceptually

Would the following be a correct description?
"The bully score for a given peer is the number of tipset snapshots that have been taken since an event from that node has been put into our ancestry. The maximum bully score is the total number of snapshots that are stored"

Reply:
added the following

```
     * <p>
     * The bully score is defined as the number times the snapshot has been advanced without updating the generation
     * of a particular node. For nodes that do not have any events that are legal other parents, the bully score
     * is defined to be 0, regardless of how many times the snapshot has been advanced.
```